02-microarray/pathway-analysis_microarray_03_gsea.Rmd

```diff
@@ -0,0 +1,318 @@
+
+Take a look at our other gene identifier conversion examples for examples with different species and gene ID types:
+[the microarray example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html) and [the RNA-seq example](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.html).
+
+The function we will use to map from Ensembl gene IDs to Entrez gene IDs is called `mapIds()`.
+
+Let's create a data frame that shows the mapped Entrez IDs along with the differential expression stats for the respective Ensembl IDs.
+
+```{r}
+# First let's create a mapped data frame we can join to the differential expression stats
+dge_mapped_df <- data.frame(
+  "entrez_id" = mapIds(
+    org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data
+    keys = dge_df$Gene,
+    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID
+  )
+) %>%
+  # If an Ensembl gene identifier doesn't map to a Entrez gene identifier, drop that
+  # from the data frame
+  dplyr::filter(!is.na(entrez_id)) %>%
+  # Make an `Ensembl` column to store the rownames
+  tibble::rownames_to_column("Ensembl") %>%
+  # Now let's join the rest of the expression data
+  dplyr::inner_join(dge_df, by = c("Ensembl" = "Gene"))
+```
+
+This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple Entrez IDs.
+In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
+For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of `dge_mapped_df`.
+
+```{r}
+head(dge_mapped_df)
+```
+
+Now let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(dge_mapped_df$entrez_id))
+```
+
+Looks like we do have duplicated Entrez IDs.
+We do not want duplicated gene identifiers for the GSEA steps later, so let's keep the Entrez IDs associated with the higher t-statistic value.
+
+```{r}
+filtered_dge_mapped_df <- dge_mapped_df %>%
+  # Sort so that highest t-statistic values are at the top
+  dplyr::arrange(dplyr::desc(t)) %>%
+  # Filter out the duplicated rows -- this will keep the first row with the
+  # duplicated value thus keeping the row with the highest t-statistic value
+  dplyr::filter(!duplicated(entrez_id))
```

Comment:
I think `distinct()` is a more direct version of the `dplyr::filter(!duplicated())` you have here. 

Second question, can we prove this to ourselves a bit? Perhaps as simply as printing out one of the duplicated entrez IDs and their t values before and after?  (I don't want to add too much length to these steps, but I also think its good to make data removal steps proved and clear).

Reply:
The reason I opted to use `dplyr::filter(!duplicated(entrez_id))` is because `dplyr::distinct(entrez_id)` returns only the column `entrez_id` while `dplyr::distinct()` returns all the rows containing duplicate identifiers (since their `t` values etc. are different). Perhaps my implementation of `dplyr::distinct()` is incorrect in this case?

Also, I agree with your second point here. While developing, I used the following to find the duplicate ids and their associated data as a sanity check:

`dge_mapped_df %>%
  dplyr::filter(duplicated(entrez_id))`

However, this returns just one of the rows with each of the duplicate ids (I manually searched the before and after data frames for the associated data using the exact `entrez_id` value returned)

<img width="946" alt="Screen Shot 2020-11-04 at 12 48 04 PM" src="https://user-images.githubusercontent.com/43576623/98150027-0a366300-1e9c-11eb-8efb-31c25824df3b.png">

Reply:
> `dplyr::distinct(entrez_id)` returns only the column `entrez_id`

Ah. There's an `.keep_all = TRUE` argument for `distinct()` that you need to use to not drop columns. 

> Also, I agree with your second point here. While developing, I used the following to find the duplicate ids and their associated data as a sanity check:
> dge_mapped_df %>% dplyr::filter(duplicated(entrez_id))

I was trying to find a straight forward way of doing this without having to make a separate duplicate entrez ids object, but I didn't come up with anything that's great. I was hoping `duplicated()` had an option to return values directly but it doesn't. I also looked to see if tidyverse had a reverse of `distinct()` but it doesn't seem to. If we installed another package (which I don't want to do) we could use `janitor::get_dupes()` but I don't find that worth  having users install another package for. 

So we are left with doing the manual preview you used here -- which I think may be the simplest route for users to follow and still get the point across. OR, this kind of thing: 
```
dup_entrez_ids <- dge_mapped_df %>% 
  dplyr::filter(duplicated(entrez_id)) %>%
  dplyr::pull(entrez_id)
```
where you then have to use `dup_entrez_ids` to retrieve things, but you'd still have to do an arrange. 

I think we should just stick with your simple and effective use of an example entrez id like `336702`.

Reply:
In the most recent commit, I went ahead and replaced the `dplyr::filter(!duplicated(entrez_id))` step with `dplyr::distinct(entrez_id, .keep_all = TRUE)` as you suggested @cansavvy, and left the subsequent steps as is because I also believe it is not worth having users install another package for. I do wish `distinct()` had a reverse function but I believe what we currently have is the next best simple yet effective solution.